### ISSUE TITLE
MultiCheckboxWidget complex form keys used incorrectly

### ISSUE THREAD

**Issue by nrother** [cakephp/cakephp#6894]
The MultiCheckboxWidget accepts two version of `options`: The simple form `['a' => 'A', 'b' => 'B']` and the complex form:
```php
['a' => [
    'value' => 'b',
    'text' => 'B'
   ]
]
```
In the first case, the array indices (a, b) are checked against `val` to find the checked items, and the indices are also used to generate the html `value`attribute. Everything fine here.

In the second case, the value (b) is used to create the html `value` attribute, but **the array index (a) is checked against  `val`**. This is a bug in my opinion, the array index should be ignored in the second case.

Context: I don't use `find('list')` to create the `options` paramter, but `find('all')` with a custom `map()` afterwards that maps the items to arrays for the second form. This works find, but the wrong items are marked as selected as the array indices are lost during this processes. (Is this the correct way to do this? Or is this just a overcomplicated way? The bug exists in any case.)

**Comment by nrother** [comment]
Ok, I just found the test for this, I'll investigate why the test works and my code not.

**Comment by nrother** [comment]
Ok, that was simple: The test does not use the complex form :smile: 

Just uncomment the line in the following snipped (and comment the over it) and the test will fail although nothing should have changed. (In fact the array index changed from 1 to 0 and this bug triggers)

```php
$data = [
            'name' => 'Tags[id]',
            'options' => [
                1 => 'CakePHP',
                //['value' => 1, 'text' => 'CakePHP'],
                '1x' => 'Development',
            ],
            'val' => [1],
            'disabled' => false
        ];
```

**Comment by markstory** [comment]
Fixed in #6895
